Module configuration variants

ycombinator opened the pull request:
Resolves #8969.

This PR introduces the concept of a module configuration variant, as detailed in #8969. This works with `filebeat` as well as `metricbeat` modules.

## Usage examples

### Listing modules and their variants

```
$ ./metricbeat modules list
Enabled:
system

Disabled:
...
docker
dropwizard
elasticsearch
elasticsearch:xpack
envoyproxy
etcd
...
```

```
$ ls -l  modules.d/elasticsearch*
-rw-r--r--  1 shaunak  staff  402 Nov 15 16:39 modules.d/elasticsearch.xpack.yml.disabled
-rw-r--r--  1 shaunak  staff  286 Nov 15 16:39 modules.d/elasticsearch.yml.disabled
```

### Enabling a module variant configuration
```
$ ./metricbeat modules enable elasticsearch:xpack
Enabled elasticsearch:xpack
```

```
$ ls -l  modules.d/elasticsearch*
-rw-r--r--  1 shaunak  staff  402 Nov 15 16:39 modules.d/elasticsearch.xpack.yml
-rw-r--r--  1 shaunak  staff  286 Nov 15 16:39 modules.d/elasticsearch.yml.disabled
```

## Source config files for a module

```
ls -l module/elasticsearch/_meta/config*
-rw-r--r--  1 shaunak  staff  377 Nov 15 11:17 module/elasticsearch/_meta/config.reference.yml
-rw-r--r--  1 shaunak  staff  276 Nov 15 11:34 module/elasticsearch/_meta/config.xpack.yml
-rw-r--r--  1 shaunak  staff  160 Oct 25 14:26 module/elasticsearch/_meta/config.yml
```

ycombinator commented:
I still need to write tests, docs and a changelog entry for the functionality in this PR but just wanted to put it out there and get some early feedback.

jsoriano commented on filebeat/Makefile:
```diff
@@ -23,8 +25,8 @@ configs: python-env
-	@rm -rf modules.d && mkdir -p modules.d
-	@for MODULE in `find module -type d -maxdepth 1 -mindepth 1 -exec basename {} \;`; do cp -a $(PWD)/module/$$MODULE/_meta/config.yml modules.d/$$MODULE.yml.disabled; done
+	@rm -rf modules.d
+	${PYTHON_ENV}/bin/python ${ES_BEATS}/script/modules_collector.py --beat ${BEAT_NAME} --docs_branch=$(DOCS_BRANCH)
```
Nice to see this in both metricbeat and filebeat :slightly_smiling_face:

jsoriano commented on script/modules_collector.py:
```diff
@@ -0,0 +1,61 @@
+import os
```
It'd be great to rewrite this as a mage target (no need to do it on this PR)

kvch commented on script/modules_collector.py:
```diff
@@ -0,0 +1,61 @@
+import os
```
AFAIK we are trying to get rid of Python scripts. I would prefer this script to be written in Golang.

jsoriano replied:
@kvch this script is not new, it has been moved from `metricbeat/scripts`, I am +1 on rewritting it in Golang, but I think this could be also done in other PR.

ycombinator replied:
Ah, I wasn't aware that we were trying to convert our build scripts to Golang. Makes sense and TIL!

So I'm also +1 for porting over this script but in a follow up PR, since this PR _mostly_ moves the existing script around to use it from filebeat as well as metricbeat. @kvch are you okay if I do this work in a follow up PR?

ycombinator commented:
@jsoriano and @kvch Thanks for your LGTMs. I've updated this PR with unit tests and CHANGELOG entries. Additionally, given that Filebeat now uses the same module config collection script as Metricbeat, the Filebeat module configuration files now each contain a header with a link to the module's documentation. Given these changes, you might want to re-review this PR.

@dedemorton I could use your guidance on where best to document the functionality introduced by this PR. Somewhere in https://www.elastic.co/guide/en/beats/filebeat/current/configuration-filebeat-modules.html seems appropriate to me but I'll defer to your opinion.

Everyone: Let's now have the naming discussion 😄. Are we okay calling these alternative out-of-the-box module configurations as "variants"? Or do folks like "flavors" better as used by @jsoriano and @ruflin in #8969? Or something else?

ruflin commented:
Could we split up this PR in two parts? 1 that moves the python files + adds new headers to config files and a second one that does the actual changes with flavors etc? Like this the flavor / variants PR will be much smaller.

For the naming: There is also a discussion around flavor for datasets. Meaning it can have different inputs but the output data structure will be identical. Here I think flavor fits well. For the module I'm asking myself if it the same thing? The x-pack one is a bit special as it creates a completely different data structure. What is a flavor of a module in a more generic case? It is just the same module with different default config options?

ycombinator commented:
> Could we split up this PR in two parts? 1 that moves the python files + adds new headers to config files and a second one that does the actual changes with flavors etc? Like this the flavor / variants PR will be much smaller.

I've made a new PR for just the python file reuse + adding new headers: #9150. And I have rebased this PR (#9118) on top of the new one (#9150). Once the new one is reviewed and merged, I will rebase this PR on `master` so it becomes smaller.

ycombinator commented:
#9150 has been merged now so I've rebased this PR on `master`. It now only contains the changes for implementing module configuration variants.

dedemorton commented:
@ycombinator 

> I could use your guidance on where best to document the functionality introduced by this PR. Somewhere in https://www.elastic.co/guide/en/beats/filebeat/current/configuration-filebeat-modules.html seems appropriate to me but I'll defer to your opinion.

That seems like the right place, but we need to make sure we show the syntax in other places where we show the enable command.

How common will the variants/flavors be? If they are likely to be fairly common (on most/all modules), maybe we should document the variant as part of the main syntax: `filebeat modules enable apache2:variant`. We'd mention that variant is optional. If no variant is specified, Filebeat uses the  the default configuration. 

I suspect that a lot users don't read the topic about specifying which modules to run. They see command examples in other places (like the getting started and the docs for each module), so we need to make sure they see the new syntax there, too. 

Re naming:

IMO, the word "variant" isn't great, but I don't have a better suggestion. I do not like flavor very much either. :-P If you compare the actual definitions of the two terms, I think "variant" is more accurate:

**variant** - a form or version of something that differs in some respect from other forms of the same thing or from a standard.

**flavor** - an indication of the essential character of something.

ycombinator commented:
> @ycombinator
> 
> > I could use your guidance on where best to document the functionality introduced by this PR. Somewhere in [elastic.co/guide/en/beats/filebeat/current/configuration-filebeat-modules.html](https://www.elastic.co/guide/en/beats/filebeat/current/configuration-filebeat-modules.html) seems appropriate to me but I'll defer to your opinion.
> 
> That seems like the right place, but we need to make sure we show the syntax in other places where we show the enable command.
> 

I've added a section on configuration variants to that page in 501c8b6. Please take a look when you get a chance.

> How common will the variants/flavors be? If they are likely to be fairly common (on most/all modules), maybe we should document the variant as part of the main syntax: `filebeat modules enable apache2:variant`. We'd mention that variant is optional. If no variant is specified, Filebeat uses the the default configuration.
> 
> I suspect that a lot users don't read the topic about specifying which modules to run. They see command examples in other places (like the getting started and the docs for each module), so we need to make sure they see the new syntax there, too.
> 

I don't expect module configuration variants to be very common, at least initially. The specific use case that prompted this feature was for X-Pack Monitoring with Metricbeat so we'll make sure to add the variant syntax to those docs. (cc: @lcawl 
 - I'll put up the PRs but just so you have context for them, this is it). Besides that, let's document the variant for a module when we introduce it, i.e. in follow up PRs to this one.

> Re naming:
> 
> IMO, the word "variant" isn't great, but I don't have a better suggestion. I do not like "flavor" very much either. :-P If you compare the actual definitions of the two terms, I think "variant" is more accurate:
> 
> **variant** - a form or version of something that differs in some respect from other forms of the same thing or from a standard.
> 
> **flavor** - an indication of the essential character of something.

Naming is hard :) Unless someone feels strongly about the name, I'm inclined to go with "variants" given your finding that it's more accurate.

@ruflin I saw your note about "flavor" above but I wasn't sure if you were advocating for or against it as compared to "variant". Putting aside the case of the xpack variant (as it's a bit special), other examples that @jsoriano [came up with](https://github.com/elastic/beats/issues/8969#issuecomment-438205136) were [`system:raid`](https://github.com/elastic/beats/blob/master/metricbeat/modules.d/system.yml#L34-L38) or [`kubernetes:events`](https://github.com/elastic/beats/blob/master/metricbeat/modules.d/kubernetes.yml.disabled#L42-L45), as these sections are completely commented out in the respective module configuration files currently.

kvch commented on libbeat/cfgfile/glob_manager.go:
```diff
@@ -182,3 +189,40 @@ func (g *GlobManager) Disable(name string) error {
 }
+
+// DisplayName returns the config file's display name
+func (f *CfgFile) DisplayName() string {
```
Why not call it `String`? I think it is more common to use functions named `String` to get the string representation of a type.

ycombinator replied:
@kvch Made this change in 9ba27c7.

ruflin commented:
What if we don't use variant or flavor but treat it as a different module? As an example: From a user perspective `kubernetes` is a module but when looking at the config it is actually a combination of multiple module configurations (or services?). Let's assume elasticsearch_xpack is it's own module, what would be our naming scheme? In this case I think we would follow the current convention with [a-z] and `_`. So this would become `elasticsearch_xpack` module. I think we need to decouple ourself on what means a module from a developer perspective and a module just form a config perspective. It's probably a bit unfortunate that we have `module: elasticsearch` in a `elasticsearch_xpack` config but we can't change that short term.

ycombinator commented:
@ruflin I'm not fully following your proposal. Using an example (say, `kubernetes` and `kubernetes_events`), can you elaborate on:

1. what folders/files would live under `module`?
2. what corresponding files would live under `module.d`?
3. how will 2. be generated from 1.?

ruflin commented:
1. The folders under `module` would stay the same but the kubernetes directory would potentially contain 2 files for the `modules.d` directory
2. The modules.d directory would contain `kuberentes.yml` and `kubernetes_events.yml`
3. I would assume manually

It's basically what you initially proposed just with a different argumentation around it.

ycombinator commented:
Chatted with @ruflin off-PR, including looking at some of his future ideas around modules configuration, and we came up with the following proposal for this PR:

* Under `module/elasticsearch/_meta/` today we have a `config.yml`. This file gets packaged as `modules.d/elasticsearch.yml` today. We will continue to do this.

* Optionally, modules may specific additional `config-VARIANT.yml` files under their `_meta/` folder, e.g. `module/elasticsearch/_meta/config-xpack.yml`. This file will get packaged as `modules.d/elasticsearch-xpack.yml`. The `-` symbol would be used to delimit module names from their variants. We are reserving `:` for future use.

ycombinator commented:
@ruflin if you're 👍 to the design in https://github.com/elastic/beats/pull/9118#issuecomment-478986292, I've updated the PR accordingly and it's ready for review now.

ruflin commented:
👍

ruflin commented on libbeat/cfgfile/glob_manager.go:
```diff
@@ -182,3 +185,31 @@ func (g *GlobManager) Disable(name string) error {
+	s[i], s[j] = s[j], s[i]
+}
+
+func (s byCfgFileDisplayNames) Less(i, j int) bool {
```
Could you add a quick test for this?

ycombinator replied:
Added in b8d0094f4.

ruflin commented on metricbeat/docs/metricbeat-options.asciidoc:
```diff
@@ -113,6 +113,20 @@ metricbeat.modules:
+one or more variant configuration files containing common alternative configurations
+for that module.
+
+When you see the list of enabled and disabled modules, those modules with configuration
```
Thanks for adding the docs 👍

ruflin commented on metricbeat/modules.d/kibana-xpack.yml.disabled:
```diff
@@ -0,0 +1,12 @@
+# Module: kibana
```
I'm a bit surprised that the collection of these files just works. Or did you move these manually?

ycombinator replied:
Me too, TBH! However, it works; I just re-tested it twice to make sure, like so:

```
rm -rf modules.d/
mage update
ls -al modules.d/ | grep xpack
```

Do you mind pulling down this PR and trying this out yourself as well, just in case?

ycombinator commented:
@ruflin Addressed all your feedback. Ready for another round of review. Thanks!

cachedout commented on CHANGELOG.next.asciidoc:
```diff
@@ -114,6 +114,7 @@ https://github.com/elastic/beats/compare/v7.0.0-alpha2...master[Check the HEAD d
 - Add AWS s3_request metricset. {pull}10949[10949] {issue}10055[10055]
 - Add s3_daily_storage metricset. {pull}10940[10940] {issue}10055[10055]
 - Add `coredns` metricbeat module. {pull}10585[10585]
+- Allow module configurations to have variants or flavors {pull}9118[9118]
```
Nitpick: this reads to me like `variants` and `flavors` are two different concepts. Since `variants` is used in the documentation you wrote, we could drop the word `flavors` here.

ruflin replied:
@ycombinator Now there seem to be 2 lines :-)

ycombinator commented:
jenkins, test this
